docs/adr/adr-11-blocksync-overhaul.md

```diff
@@ -0,0 +1,372 @@
+engine for managing any CAR files with indexing, convenient abstractions, tools, recovery mechanisms, etc.
+  - __EDSes as _CARv1_ files over _CARv2_.__ CARv2 encodes indexes into the file, while DAGStore maintains CARv2-based
+indexing. Usage of CARv1 keeps only one copy of the index, stores/transfers less metadata per EDS, and simplifies
+reading EDS from a file.
+
+- __LNs DASing remains untouched__. The networking protocol and storage for LNs remain intact as it fulfills the
+requirements. Bitswap usage as the backbone protocol for requesting samples and global Badger KVStore remain unaffected.
+
+### Detailed Design
+
+> All the comments on the API definitions should be preserved and potentially improved by implementations.
+
+The new block storage design is solely additive. All the existing storage-related components and functionality
+are kept with additional components introduced. Altogether, existing and new components will be recomposed to serve as the
+foundation of our improved block storage subsystem.
+
+The central data structure representing Celestia block data is EDS(`rsmt2d.ExtendedDataSquare`), and the new storage design
+is focused around storing entire EDSes as a whole rather than a set of individual chunks, s.t. storage subsystem
+can handle storing and streaming/serving blocks of 4MB and more.
+
+#### EDS Serde
+
+Storing EDS as a whole requires EDS (de)serialization. For this, the [CAR format](https://ipld.io/specs/transport/car) is
+chosen.
+
+##### `share.WriteEDS`
+
+To write EDS into a stream/file, `WriteEDS` is introduced. Internally, it
+
+- [Re-imports](https://github.com/celestiaorg/rsmt2d/blob/master/extendeddatasquare.go#L41) EDS similarly to
+[`ipld.ImportShares`](https://github.com/celestiaorg/celestia-node/blob/main/ipld/add.go#L51-L65)
+  - Using [`Blockservice`](https://github.com/ipfs/go-blockservice/blob/master/blockservice.go#L46) with [offline
+exchange](https://github.com/ipfs/go-ipfs-exchange-offline/blob/master/offline.go#L16) and in-memory [`Blockstore`](https://github.com/ipfs/go-ipfs-blockstore/blob/master/blockstore.go#L33)
+  - With [`NodeVisitor`](https://github.com/celestiaorg/celestia-node/blob/main/ipld/add.go#L56), which saves to the
+`Blockstore` only NMT Merkle proofs(no shares) _NOTE: `len(node.Links()) == 2`_
+    - Actual shares are written further in a particular way explained further
+- Creates and [writes](https://github.com/ipld/go-car/blob/master/car.go#L86) header [`CARv1Header`](https://github.com/ipld/go-car/blob/master/car.go#L30)
+  - Fills up `Roots` field with `EDS.RowRoots/EDS.ColRoots` roots converted into CIDs
+- Iterates over shares in quadrant-by-quadrant order via `EDS.GetCell`
+  - [Writes](https://github.com/ipld/go-car/blob/master/car.go#L118) the shares in row-by-row order
+- Iterates over in-memory Blockstore and [writes]((https://github.com/ipld/go-car/blob/master/car.go#L118)) NMT Merkle
+proofs stored in it
+
```

Comment:
All above links should be perma-links. Otherwise they will be out of date shortly.